### ISSUE TITLE
Feature Request: support Dubbo RPC protocol

### ISSUE THREAD

**Issue by yun1129** [envoyproxy/envoy#3998]
Dubbo community decided to make contribution to Envoy by adding Dubbo RPC protocol support. Open this issue for tracking the work.

A draft design document is available at below link. https://docs.google.com/document/d/1A8xXqWJ0kIMmPNBKiGjLBTN-4FLarxMiyylYv8Dwpe8/edit?usp=sharing

Thanks for your review and comments.

**Comment by mattklein123** [comment]
Thanks @yun1129. Assigning to myself to review the proposal. @zuercher if you feel like also taking a look that would be cool since I suspect there is going to be some overlap with Thrift.

**Comment by mattklein123** [comment]
@yun1129 doc looks great. Very excited to help with reviews for this!

**Comment by yun1129** [comment]
@mattklein123, the code is almost ready and I would like to seek your suggestion for upcoming steps. However, there is one important thing we should address - the integration test.

From Dubbo community perspective there are only samples for illustrating how Dubbo RPC framework works and there is no integration test environment. Furthermore, those samples are written in Java. From Envoy perspective I believe integration test is crucial but we need a simple and light solution for fulfilling that rather than bringing the Java runtime into Envoy's integration test environment.

I'm talking to Dubbo community for this. Meanwhile would like to hear your voice.

**Comment by mattklein123** [comment]
@yun1129 it's not a perfect integration test, but since you have to write both a decoder and encoder for proxy, I would probably just implement a fake upstream server in C++ using the Envoy encode/decoder code. This should be relatively simple to do and give good coverage. Obviously, it won't catch compat issues in the encoder/decoder itself, but since I'm guessing the protocol is pretty stable I think some manual testing is probably good enough and the built-in C++ integration test server can catch major regressions. Thoughts?

**Comment by yun1129** [comment]
@mattklein123, in the long run, I believe you suggested solution is the final solution. Currently, Dubbo community doesn't have a light solution, meaning the upstream server is written in C++ or Go, that can fit this purpose. That's to say my team has to spend time to implement it and it leads time intensive. However, a Dubbo meetup is scheduled for the end of this month we don't have enough time to get the integration test ready and push the code of phase into Envoy repository. Ideally, Dubbo community would like to get the code open source when it announces Dubbo Mesh.

An alternative way is forking Envoy into the Dubbo community and once the work for integration test and phase two is done we can PR it back to Envoy.

Appreciate your suggestions!

**Comment by mattklein123** [comment]
@yun1129 sorry I didn't mean for you to write a production quality C++ Dubbo server. All I meant was that in order to implement the Dubbo proxy/router, you will need both an encoder and a decoder. With those, it should be pretty easy to make a FakeDubboServer using our integration test primitives. For example in this test (https://github.com/envoyproxy/envoy/blob/master/test/integration/ratelimit_integration_test.cc) we make a fake gRPC rate limit server and run some tests with that.

With that said, we don't have any integration tests for Redis (although arguably a simpler protocol) so if you want to start with just unit tests that's fine too.

**Comment by yun1129** [comment]
@mattklein123 if it's acceptable I would like to take unit tests as the start for quality assurance. Since the final goal is making Dubbo Mesh can be utilized on production and my team will spend time on integration test for building quality confidence up.

**Comment by yun1129** [comment]
Sure, no problem.

**Comment by lizan** [comment]
@zyfjeff is working on this at #4200

**Comment by mattklein123** [comment]
@zyfjeff thoughts on adding some docs around this new filter? I think then we can call this initial issue done?

**Comment by zyfjeff** [comment]
@mattklein123 
I am going to add documentation next week, but this issue can continue to be kept because we are doing Dubbo routing. 

ref: #5251

**Comment by NameHaibinZhang** [comment]
this pr is already merge in envoy?

**Comment by zyfjeff** [comment]
@NameHaibinZhang Has been merged into the master branch

**Comment by hzxuzhonghu** [comment]
Is Dubbo Support complete now ?

**Comment by mattklein123** [comment]
This is done. Let's track new feature request/issues with new issues.

**Comment by zhongfox** [comment]
> Is Dubbo Support complete now ?

@hzxuzhonghu any roadmap about dubbo in istio project？
